shim.c

```diff
@@ -1831,6 +1831,73 @@ debug_hook(void)
 	x = 1;
 }
 
+EFI_STATUS
+check_shim_sbat()
+{
+	unsigned int i;
+	EFI_STATUS efi_status;
+	list_t sbat_var_entries;
+	struct sbat sbat = { 0 };
+	struct sbat_entry *entry = NULL;
+	char *sbat_start = (char *)&_sbat;
+	char *sbat_end = (char *)&_esbat;
+	size_t sbat_size = sbat_end - sbat_start;
+
+	INIT_LIST_HEAD(&sbat_var_entries);
+
+	char *sbat_data = AllocatePool(sbat_size + 1);
+	if (!sbat_data) {
+		perror(L"Failed to allocate SBAT buffer\n");
+		return EFI_OUT_OF_RESOURCES;
+	}
+
+	CopyMem(sbat_data, sbat_start, sbat_size);
+	sbat_data[sbat_size] = '\0';
+
+	efi_status = parse_sbat(sbat_data, sbat_size + 1, &sbat);
+	if (EFI_ERROR(efi_status)) {
+		perror(L"shim SBAT data not correct: %r\n", efi_status);
+		goto out;
+	}
+
+	dprint(L"shim SBAT data\n");
+	for (i = 0; i < sbat.size; i++) {
+		entry = sbat.entries[i];
+		dprint(L"%a, %a, %a, %a, %a, %a\n",
+			entry->component_name,
+			entry->component_generation, entry->vendor_name,
+			entry->vendor_package_name, entry->vendor_version,
+			entry->vendor_url);
+	}
+
+	efi_status = parse_sbat_var(&sbat_var_entries);
+	/*
+	 * If the SBAT variable is not present, the
+	 * .sbat section doesn't have to be verified.
+	 */
+	if (efi_status == EFI_NOT_FOUND) {
+		efi_status = EFI_SUCCESS;
+		goto out;
+	} else if (EFI_ERROR(efi_status)) {
+		perror(L"Parsing SBAT variable failed: %r\n", efi_status);
+		goto out;
+	}
+
+	efi_status = verify_sbat(&sbat, &sbat_var_entries);
+	if (EFI_ERROR(efi_status))
+		perror(L"shim SBAT verification failed: %r\n", efi_status);
+out:
+	if (sbat.entries) {
+		for (i = 0; i < sbat.size; i++)
+			FreePool(sbat.entries[i]);
+		FreePool(sbat.entries);
+	}
+	FreePool(sbat_data);
+	cleanup_sbat_var(&sbat_var_entries);
```

Comment:
Maybe we could just add cleanup to verify_sbat failures, and you won't need it at all. Though up to you to decide, less stuff for me to change ) I believe we have:
1) cleanup on parse_sbat failure ( internal )
2) cleanup on verify_sbat success

if we add cleanup on verify_sbat failure we should not need it outside, do we ? or we have a condition when we need cleanup in-between parse and verify ?